container/go/pkg/compat/image.go

```diff
@@ -0,0 +1,147 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//////////////////////////////////////////////////////////////////////
+// Image for intermediate format used in python containerregistry.
+// Uses the go-containerregistry API as backend.
+
+package compat
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"sync"
+
+	v1 "github.com/google/go-containerregistry/pkg/v1"
+	"github.com/google/go-containerregistry/pkg/v1/partial"
+	"github.com/google/go-containerregistry/pkg/v1/types"
+)
+
+// This intermediate layout image implements v1.Image, its implementation is very similar to layout.layoutImage.
+type layoutImage struct {
+	// path is the path of this layout, with helper functions for finding the full directory.
+	path Path
+	// desc is the descriptor metadata for this image.
+	desc v1.Descriptor
+	// manifestLock protects rawManifest.
+	manifestLock sync.Mutex
+	// rawManifest is the raw bytes of manifest.json file.
+	rawManifest []byte
+}
+
+var _ partial.CompressedImageCore = (*layoutImage)(nil)
+
+// MediaType of this image's manifest.
+func (li *layoutImage) MediaType() (types.MediaType, error) {
+	return li.desc.MediaType, nil
+}
+
+// Parses manifest.json into Manifest object. Implements WithManifest for partial.Blobset.
+func (li *layoutImage) Manifest() (*v1.Manifest, error) {
+	return partial.Manifest(li)
+}
+
+// RawManifest returns the serialized bytes of manifest.json metadata.
+func (li *layoutImage) RawManifest() ([]byte, error) {
+	li.manifestLock.Lock()
+	defer li.manifestLock.Unlock()
+	if li.rawManifest != nil {
+		return li.rawManifest, nil
+	}
+
+	b, err := ioutil.ReadFile(li.path.path(manifestFile))
+	if err != nil {
+		return nil, err
+	}
+
+	li.rawManifest = b
+	return li.rawManifest, nil
+}
+
+// RawConfigFile returns the serialized bytes of config.json metadata.
+func (li *layoutImage) RawConfigFile() ([]byte, error) {
+	return ioutil.ReadFile(li.path.path(configFile))
+}
+
+// LayerByDigest returns a Layer for interacting with a particular layer of the image, looking it up by "digest" (the compressed hash).
+// We assume the layer files are named in the format of e.g., 000.tar.gz in this path, following the order they appear in manifest.json.
+func (li *layoutImage) LayerByDigest(h v1.Hash) (partial.CompressedLayer, error) {
+	manifest, err := li.Manifest()
+	if err != nil {
+		return nil, err
+	}
+
+	// The config is a layer in some cases.
+	if h == manifest.Config.Digest {
+		return partial.CompressedLayer(&compressedBlob{
+			path:  li.path,
+			desc:  manifest.Config,
+			index: -1,
+		}), nil
+	}
+
+	for i, desc := range manifest.Layers {
+		if h == desc.Digest {
+			switch desc.MediaType {
+			case types.OCILayer, types.DockerLayer:
+				return partial.CompressedToLayer(&compressedBlob{
+					path: li.path,
+					desc: desc,
+					// Passed in index to look for the actual gzipped layer file due to intermediate format naming convention.
+					index: i,
```

Comment:
instead of setting the index, I would recommend storing the name of the file here, ie layerFilename(i). This way the implementation of the `Compressed` function on `compressedBlob` won't need the control statement.